In [3]:
Attribute_Information_dict = {
    1:  ['surgery?',
          '1 = Yes, it had surgery',
          '2 = It was treated without surgery'],
    2: ['Age',
          '1 = Adult horse',
          '2 = Young (< 6 months)'],
    3: ['Hospital Number',
          '- numeric id',
          '- the case number assigned to the horse',
            '(may not be unique if the horse is treated > 1 time)'],
    4: ['Rectal temperature',
          '- linear',
          '- in degrees celsius.',
          '- An elevated temp may occur due to infection.',
          '- temperature may be reduced when the animal is in late shock',
          '- normal temp is 37.8',
          '- this parameter will usually change as the problem progresses',
               'eg. may start out normal, then become elevated because of',
                   'the lesion, passing back through the normal range as the',
                   'horse goes into shock'],
    5: ['pulse',
          '- linear',
          '- the heart rate in beats per minute',
          '- is a reflection of the heart condition: 30 -40 is normal for adults',
          '- rare to have a lower than normal rate although athletic horses',
            'may have a rate of 20-25',
          '- animals with painful lesions or suffering from circulatory shock',
            'may have an elevated heart rate'],
    6: ['respiratory rate',
          '- linear',
          '- normal rate is 8 to 10',
          '- usefulness is doubtful due to the great fluctuations'],
    7: ['temperature of extremities',
          '- a subjective indication of peripheral circulation',
          '- possible values:',
               '1 = Normal',
               '2 = Warm',
               '3 = Cool',
               '4 = Cold',
          '- cool to cold extremities indicate possible shock',
          '- hot extremities should correlate with an elevated rectal temp.'],
    8: ['peripheral pulse',
          '- subjective',
          '- possible values are:',
               '1 = normal',
               '2 = increased',
               '3 = reduced',
               '4 = absent',
          '- normal or increased p.p. are indicative of adequate circulation',
            'while reduced or absent indicate poor perfusion'],
    9: ['mucous membranes',
          '- a subjective measurement of colour',
          '- possible values are:',
               '1 = normal pink',
               '2 = bright pink',
               '3 = pale pink',
               '4 = pale cyanotic',
               '5 = bright red / injected',
               '6 = dark cyanotic',
          '- 1 and 2 probably indicate a normal or slightly increased',
            'circulation',
          '- 3 may occur in early shock',
          '- 4 and 6 are indicative of serious circulatory compromise',
          '- 5 is more indicative of a septicemia'],
    10: ['capillary refill time',
          '- a clinical judgement. The longer the refill, the poorer the circulation',
           '- possible values',
               '1 = < 3 seconds',
               '2 = >= 3 seconds'],
    11: ['pain - a subjective judgement of the horse"s pain level',
          '- possible values:',
               '1 = alert, no pain',
               '2 = depressed',
               '3 = intermittent mild pain',
               '4 = intermittent severe pain',
               '5 = continuous severe pain',
          '- should NOT be treated as a ordered or discrete variable!',
          '- In general, the more painful, the more likely it is to require surgery',
          '- prior treatment of pain may mask the pain level to some extent'],
    12: ['peristalsis',
          '- an indication of the activity in the horse"s gut. As the gut',
            'becomes more distended or the horse becomes more toxic, the',
            'activity decreases',
          '- possible values:',
               '1 = hypermotile',
               '2 = normal',
               '3 = hypomotile',
               '4 = absent'],
    13: ['abdominal distension',
          '- An IMPORTANT parameter.',
          '- possible values',
               '1 = none',
               '2 = slight',
               '3 = moderate',
               '4 = severe',
          '- an animal with abdominal distension is likely to be painful and',
            'have reduced gut motility.',
          '- a horse with severe abdominal distension is likely to require',
            'surgery just to relieve the pressure'],
    14: ['nasogastric tube',
          '- this refers to any gas coming out of the tube',
          '- possible values:',
               '1 = none',
               '2 = slight',
               '3 = significant',
          '- a large gas cap in the stomach is likely to give the horse discomfort'],
    15: ['nasogastric reflux',
          '- possible values',
               '1 = none',
               '2 = > 1 liter',
               '3 = < 1 liter',
          '- the greater amount of reflux, the more likelihood that there is',
            'some serious obstruction to the fluid passage from the rest of the intestine'],
    16: ['nasogastric reflux PH',
          '- linear',
          '- scale is from 0 to 14 with 7 being neutral',
          '- normal values are in the 3 to 4 range'],
    17: ['rectal examination - feces',
          '- possible values',
               '1 = normal',
               '2 = increased',
               '3 = decreased',
               '4 = absent',
          '- absent feces probably indicates an obstruction'],
    18: ['abdomen',
          '- possible values',
               '1 = normal',
               '2 = other',
               '3 = firm feces in the large intestine',
               '4 = distended small intestine',
               '5 = distended large intestine',
          '- 3 is probably an obstruction caused by a mechanical impaction and is normally treated medically',
          '- 4 and 5 indicate a surgical lesion'],
    19: ['packed cell volume',
          '- linear',
          '- the # of red cells by volume in the blood',
          '- normal range is 30 to 50. The level rises as the circulation',
            'becomes compromised or as the animal becomes dehydrated.'],
    20: ['total protein',
          '- linear',
          '- normal values lie in the 6-7.5 (gms/dL) range',
          '- the higher the value the greater the dehydration'],
    21: ['abdominocentesis appearance',
          '- a needle is put in the horse"s abdomen and fluid is obtained from',
            'the abdominal cavity',
          '- possible values:',
               '1 = clear',
               '2 = cloudy',
               '3 = serosanguinous',
          '- normal fluid is clear while cloudy or serosanguinous indicates a compromised gut'],
    22: ['abdomcentesis total protein',
          '- linear',
          '- the higher the level of protein the more likely it is to have a',
            'compromised gut. Values are in gms/dL'],
    23: ['outcome',
          '- what eventually happened to the horse?',
          '- possible values:',
               '1 = lived',
               '2 = died',
               '3 = was euthanized'],
    24: ['surgical lesion?',
          '- retrospectively, was the problem (lesion) surgical?',
          '- all cases are either operated upon or autopsied so that',
           ' this value and the lesion type are always known',
          '- possible values:',
               '1 = Yes',
               '2 = No'],
    25: ['type of lesion',
          '- first number is site of lesion',
               '1 = gastric',
               '2 = sm intestine',
               '3 = lg colon',
               '4 = lg colon and cecum',
               '5 = cecum',
               '6 = transverse colon',
               '7 = retum/descending colon',
               '8 = uterus',
               '9 = bladder',
               '11 = all intestinal sites',
               '00 = none',
          '- second number is type',
               '1 = simple',
               '2 = strangulation',
               '3 = inflammation',
               '4 = other',
          '- third number is subtype',
               '1 = mechanical',
               '2 = paralytic',
               '0 = n/a',
          '- fourth number is specific code',
               '1 = obturation',
               '2 = intrinsic',
               '3 = extrinsic',
               '4 = adynamic',
               '5 = volvulus/torsion',
               '6 = intussuption',
               '7 = thromboembolic',
               '8 = hernia',
               '9 = lipoma/slenic incarceration',
               '10 = displacement',
               '0 = n/a'],
    26: ['check 25'],
    27: ['check 25'],
    28: ['cp_data',
          '- is pathology data present for this case?',
               '1 = Yes',
               '2 = No',
          '- this variable is of no significance since pathology data',
            'is not included or collected for these cases']
}

In [4]:
import pandas as pd
import numpy as np

### Читаем файл и задаём имена колонкам как атрибуты

In [5]:
names = list(range(1,29)) 

In [6]:
horse_data_df = pd.read_csv('horse_data.csv', names=names)
horse_data_df

1   2        3      4    5   6  7  8  9  10  ...     19    20 21    22 23  \
0    2   1   530101  38.50   66  28  3  3  ?  2  ...  45.00  8.40  ?     ?  2   
1    1   1   534817   39.2   88  20  ?  ?  4  1  ...     50    85  2     2  3   
2    2   1   530334  38.30   40  24  1  1  3  1  ...  33.00  6.70  ?     ?  1   
3    1   9  5290409  39.10  164  84  4  1  6  2  ...  48.00  7.20  3  5.30  2   
4    2   1   530255  37.30  104  35  ?  ?  6  2  ...  74.00  7.40  ?     ?  2   
..  ..  ..      ...    ...  ...  .. .. .. .. ..  ...    ...   ... ..   ... ..   
295  1   1   533886      ?  120  70  4  ?  4  2  ...     55    65  ?     ?  3   
296  2   1   527702  37.20   72  24  3  2  4  2  ...  44.00     ?  3  3.30  3   
297  1   1   529386  37.50   72  30  4  3  4  1  ...  60.00  6.80  ?     ?  2   
298  1   1   530612  36.50  100  24  3  3  3  1  ...  50.00  6.00  3  3.40  1   
299  1   1   534618   37.2   40  20  ?  ?  ?  ?  ...     36    62  1     1  3   

    24     25 26 27 28  
0    2  11300  0  0  2  
1    2   2208  0  0  2  
2    2      0  0  0  1  
3    1   2208  0  0  1  
4    2   4300  0  0  2  
..  ..    ... .. .. ..  
295  2   3205  0  0  2  
296  1   2208  0  0  1  
297  1   3205  0  0  2  
298  1   2208  0  0  1  
299  2   6112  0  0  2  

[300 rows x 28 columns]

In [7]:
df_small = horse_data_df[[1, 4, 5, 12, 13, 18, 23, 24]]
df_small

1      4    5  12 13 18 23  24
0    2  38.50   66  4  4  5  2   2
1    1   39.2   88  4  2  2  3   2
2    2  38.30   40  3  1  1  1   2
3    1  39.10  164  4  4  ?  2   1
4    2  37.30  104  ?  ?  ?  2   2
..  ..    ...  ... .. .. .. ..  ..
295  1      ?  120  4  ?  5  3   2
296  2  37.20   72  3  3  4  3   1
297  1  37.50   72  4  3  5  2   1
298  1  36.50  100  3  3  4  1   1
299  1   37.2   40  ?  ?  1  3   2

[300 rows x 8 columns]

In [8]:
df_small = df_small.replace('?', 0).astype(float) # для удобства пока так

### 1. 1 - surgery?

In [9]:
Attribute_Information_dict[1]

['surgery?', '1 = Yes, it had surgery', '2 = It was treated without surgery']

In [10]:
df_small[1].value_counts()

1.0    180
2.0    119
0.0      1
Name: 1, dtype: int64

#### В столбце surgery видим один пропуск - это меньше процента, добавим к 119, для ровного счёта.

In [11]:
df_small.loc[df_small[1] == 0.0, [1]] = 2

In [12]:
df_small[1].value_counts()

1.0    180
2.0    120
Name: 1, dtype: int64

In [13]:
print('Процентное соотношение прооперированных к общему числу осмотренных:', 180 / 300 * 100, '%')

Процентное соотношение прооперированных к общему числу осмотренных: 60.0 %


### 2. 4 - Rectal temperature

In [14]:
Attribute_Information_dict[4]

['Rectal temperature',
 '- linear',
 '- in degrees celsius.',
 '- An elevated temp may occur due to infection.',
 '- temperature may be reduced when the animal is in late shock',
 '- normal temp is 37.8',
 '- this parameter will usually change as the problem progresses',
 'eg. may start out normal, then become elevated because of',
 'the lesion, passing back through the normal range as the',
 'horse goes into shock']

#### В столбце Rectal temperature видим много пропусков, можно их выделить и изучить отдельно

In [15]:
df_small[4].round().value_counts()

38.0    152
0.0      60
39.0     47
37.0     24
40.0     10
36.0      5
41.0      1
35.0      1
Name: 4, dtype: int64

#### Изучая этот маленький фрейм видим, что у этих 4-х с температурой выше 40: повышенный пульс, замедлена работа кишечника (12)

In [16]:
df_small[df_small[4] > 39.9] # видимо, начиная с 40 градусов и выше не выживают

1     4      5    12   13   18   23   24
54   2.0  40.3  114.0  3.0  3.0  5.0  3.0  1.0
91   2.0  40.3  114.0  3.0  3.0  5.0  2.0  1.0
259  1.0  40.8   72.0  3.0  1.0  0.0  2.0  1.0
281  2.0  40.0   78.0  3.0  1.0  1.0  2.0  1.0

In [17]:
# Ок, тогда для тех, у кого нет данных о температуре и их постиг летальный исход:
# принимаем повышенную температуру 40.23 (ссылка на столбец 23)

df_small.loc[(df_small[4] == 0) & (df_small[23].isin([2, 3])), [4]] = 40.23

#### Известно, что нарушение работы кишечника с большой вероятностью может стать причиной интоксикации организма и привести к повышению температуры.

In [18]:
# Находим тех, у кого нет данных о температуре и работа пищеварительной системы замедлена или отсутствует.
    
df_small[(df_small[4] == 0) & ((df_small[12] == 3)|(df_small[12] == 4))]

1    4      5    12   13   18   23   24
5    2.0  0.0    0.0  3.0  2.0  3.0  1.0  2.0
34   1.0  0.0  100.0  4.0  4.0  4.0  1.0  1.0
52   2.0  0.0    0.0  3.0  1.0  2.0  1.0  2.0
65   1.0  0.0   60.0  3.0  2.0  4.0  1.0  1.0
96   2.0  0.0   84.0  3.0  1.0  2.0  1.0  1.0
105  1.0  0.0   70.0  3.0  2.0  5.0  1.0  1.0
112  1.0  0.0   40.0  3.0  1.0  5.0  1.0  1.0
115  2.0  0.0    0.0  3.0  2.0  0.0  1.0  2.0
145  2.0  0.0   80.0  3.0  3.0  0.0  1.0  1.0
152  1.0  0.0  100.0  4.0  4.0  0.0  1.0  1.0
167  2.0  0.0   44.0  3.0  1.0  1.0  1.0  2.0
178  1.0  0.0   60.0  4.0  4.0  0.0  1.0  1.0
219  1.0  0.0   52.0  4.0  3.0  4.0  1.0  1.0
222  1.0  0.0  112.0  3.0  1.0  4.0  1.0  1.0
253  2.0  0.0  112.0  4.0  2.0  0.0  1.0  2.0
254  1.0  0.0   80.0  4.0  4.0  5.0  1.0  1.0
272  1.0  0.0  100.0  4.0  4.0  4.0  1.0  1.0

In [19]:
# И так, тем, у кого нет данных о температуре, работа пищеварительной системы замедлена или отсутствует:
# дадим 38.12 - умеренно повышенную со ссылкой на столбец 12

df_small.loc[(df_small[4] == 0.0) & ((df_small[12] == 3)|(df_small[12] == 4)), [4]] = 38.12

In [20]:
# Остаётся вот что
df_small[df_small[4] == 0]

1    4      5    12   13   18   23   24
56   1.0  0.0    0.0  0.0  0.0  0.0  1.0  1.0
68   2.0  0.0   52.0  0.0  0.0  0.0  1.0  2.0
159  2.0  0.0    0.0  1.0  2.0  3.0  1.0  2.0
160  1.0  0.0    0.0  0.0  0.0  5.0  1.0  1.0
175  1.0  0.0    0.0  0.0  0.0  0.0  1.0  1.0
191  2.0  0.0  100.0  1.0  1.0  0.0  1.0  2.0
216  2.0  0.0    0.0  0.0  0.0  0.0  1.0  2.0
227  1.0  0.0    0.0  0.0  0.0  0.0  1.0  1.0
288  1.0  0.0    0.0  0.0  0.0  0.0  1.0  1.0

In [21]:
# Выбросим строчки с пропусками данных 
df_small = df_small.replace(0, np.NaN).dropna(thresh=5)
df_small = df_small.replace(np.NaN, 0)
df_small

1      4      5    12   13   18   23   24
0    2.0  38.50   66.0  4.0  4.0  5.0  2.0  2.0
1    1.0  39.20   88.0  4.0  2.0  2.0  3.0  2.0
2    2.0  38.30   40.0  3.0  1.0  1.0  1.0  2.0
3    1.0  39.10  164.0  4.0  4.0  0.0  2.0  1.0
4    2.0  37.30  104.0  0.0  0.0  0.0  2.0  2.0
..   ...    ...    ...  ...  ...  ...  ...  ...
295  1.0  40.23  120.0  4.0  0.0  5.0  3.0  2.0
296  2.0  37.20   72.0  3.0  3.0  4.0  3.0  1.0
297  1.0  37.50   72.0  4.0  3.0  5.0  2.0  1.0
298  1.0  36.50  100.0  3.0  3.0  4.0  1.0  1.0
299  1.0  37.20   40.0  0.0  0.0  1.0  3.0  2.0

[290 rows x 8 columns]

In [22]:
df_small[df_small[4] == 0]

1    4      5    12   13   18   23   24
159  2.0  0.0    0.0  1.0  2.0  3.0  1.0  2.0
191  2.0  0.0  100.0  1.0  1.0  0.0  1.0  2.0

In [23]:
# Оставшиеся заполним 37.77 - близко к норме, но если попадутся на глаза будем понимать, что это предположение
df_small.loc[df_small[4] == 0, [4]] = 37.77

In [24]:
df_small[4].value_counts().head(10)

40.23    31
38.00    26
38.50    19
38.30    18
37.80    17
38.12    17
38.20    16
38.60    12
37.50    12
38.10    12
Name: 4, dtype: int64

In [25]:
df_small[4].max()

40.8

In [26]:
df_small[4].astype(float).min()

35.4

In [27]:
df_small[4].astype(float).round().mode() # округляем до целых чисел и видим, что мода на норме

0    38.0
dtype: float64

In [28]:
df_small[4].median()

38.2

#### Принимаем, что выбросы по температуре не встретились.

### 3. 5 - pulse

In [29]:
Attribute_Information_dict[5]

['pulse',
 '- linear',
 '- the heart rate in beats per minute',
 '- is a reflection of the heart condition: 30 -40 is normal for adults',
 '- rare to have a lower than normal rate although athletic horses',
 'may have a rate of 20-25',
 '- animals with painful lesions or suffering from circulatory shock',
 'may have an elevated heart rate']

#### По аналогии с температурой попробуем определить смертельный порог

In [30]:
df_small[df_small[5] > 110][23].value_counts()

2.0    13
1.0    11
3.0     7
Name: 23, dtype: int64

In [31]:
# У кого нет данных и их постиг летальный исход: принимаем 120
df_small.loc[(df_small[5] == 0) & (df_small[23].isin([2,3])), [5]] = 120

In [32]:
# Остались выжившие и видно, что для большинства из низ дело обошлось без операции
df_small[df_small[5] == 0]

1      4    5    12   13   18   23   24
5    2.0  38.12  0.0  3.0  2.0  3.0  1.0  2.0
52   2.0  38.12  0.0  3.0  1.0  2.0  1.0  2.0
115  2.0  38.12  0.0  3.0  2.0  0.0  1.0  2.0
126  1.0  38.70  0.0  4.0  2.0  0.0  1.0  1.0
159  2.0  37.77  0.0  1.0  2.0  3.0  1.0  2.0
204  1.0  39.20  0.0  4.0  2.0  0.0  1.0  1.0

In [33]:
# Заполним модой
df_small.loc[df_small[5] == 0, [5]] = df_small[5].mode()

In [34]:
df_small[5].median()

65.5

In [35]:
df_small[5].max()

184.0

In [36]:
df_small[5].min()

30.0

In [37]:
print('Пульс в пределах нормы у', len(df_small[(df_small[5] > 25) & (df_small[5] < 45)][5]))

Пульс в пределах нормы у 45


In [38]:
print('Пульс выше нормы у', len(df_small[(df_small[5] > 50)][5]))

Пульс выше нормы у 201


In [39]:
print('Медиана по завышенным:', df_small[(df_small[5] > 50)][5].astype(int).median(), 'ударов в минуту.')

Медиана по завышенным: 80.0 ударов в минуту.


In [40]:
print('Мода по завышенным:', df_small[(df_small[5] > 50)][5].mode()[0], 'ударов в минуту.')

Мода по завышенным: 60.0 ударов в минуту.


#### Принимаем, что здесь так же обошлось без выбросов (184 выглядит сильно высоким значением пульса для нормы в 30-40 ударов, но видно, как остальные значения плавно подбираются к этому максимуму: 140 - 150 - 160 - 164 - 184)

### 5. 12 - peristalsis

In [41]:
Attribute_Information_dict[12]

['peristalsis',
 '- an indication of the activity in the horse"s gut. As the gut',
 'becomes more distended or the horse becomes more toxic, the',
 'activity decreases',
 '- possible values:',
 '1 = hypermotile',
 '2 = normal',
 '3 = hypomotile',
 '4 = absent']

In [42]:
df_small[12].value_counts()

3.0    128
4.0     73
1.0     39
0.0     34
2.0     16
Name: 12, dtype: int64

In [43]:
# Видим отношение замедленной работы кишечника (плюс отсутствие активности вообще) ко всем
print('Работа кишечника замедлена или отсутствует:', round(len(df_small.loc[df_small[12].isin([3, 4])]) / len(df_small), 2) * 100,'%')

Работа кишечника замедлена или отсутствует: 69.0 %


Пропуски не заполняем. Лучше вообще не иметь данных об этом атрибуте в каком-то конкретном случае, чем искажать статистику в целом.

### 6. 13 - abdominal distension

In [44]:
Attribute_Information_dict[13]

['abdominal distension',
 '- An IMPORTANT parameter.',
 '- possible values',
 '1 = none',
 '2 = slight',
 '3 = moderate',
 '4 = severe',
 '- an animal with abdominal distension is likely to be painful and',
 'have reduced gut motility.',
 '- a horse with severe abdominal distension is likely to require',
 'surgery just to relieve the pressure']

In [45]:
df_small[13].value_counts()

1.0    76
3.0    65
2.0    65
0.0    46
4.0    38
Name: 13, dtype: int64

In [46]:
df_small.loc[df_small[13].isin([3,4])][1].value_counts()

1.0    77
2.0    26
Name: 1, dtype: int64

Тут можем увидеть, что тем, кто со вздутием чаще делают операцию (>70%) и 60% погибают. Эта информация может быть полезной для спасения поступающих животных в будущем.

### 4. 18 - abdomen

In [47]:
Attribute_Information_dict[18]

['abdomen',
 '- possible values',
 '1 = normal',
 '2 = other',
 '3 = firm feces in the large intestine',
 '4 = distended small intestine',
 '5 = distended large intestine',
 '- 3 is probably an obstruction caused by a mechanical impaction and is normally treated medically',
 '- 4 and 5 indicate a surgical lesion']

In [48]:
df_small[18].value_counts()

0.0    109
5.0     78
4.0     43
1.0     28
2.0     19
3.0     13
Name: 18, dtype: int64

Виден большой процент с хирургическим повреждением (4, 5): 41%.

И ещё много пропусков - просто игнорируем их.

### 7. 23 - Исход

In [49]:
Attribute_Information_dict[23]

['outcome',
 '- what eventually happened to the horse?',
 '- possible values:',
 '1 = lived',
 '2 = died',
 '3 = was euthanized']

In [50]:
df_small[23].value_counts()

1.0    171
2.0     74
3.0     44
0.0      1
Name: 23, dtype: int64

### 8. 24 - surgical lesion?

In [51]:
Attribute_Information_dict[24]

['surgical lesion?',
 '- retrospectively, was the problem (lesion) surgical?',
 '- all cases are either operated upon or autopsied so that',
 ' this value and the lesion type are always known',
 '- possible values:',
 '1 = Yes',
 '2 = No']

In [52]:
df_small[24].value_counts()

1.0    184
2.0    106
Name: 24, dtype: int64